src/Simulation/Simulators/OpenSystemsSimulator/NativeInterface.cs

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Text.Json;
+using System.Runtime.InteropServices;
+using Microsoft.Quantum.Simulation.Core;
+using Newtonsoft.Json.Linq;
+
+#nullable enable
+
+namespace Microsoft.Quantum.Experimental
+{
+    /// <summary>
+    ///     Represents an exception that is raised by native simulator code.
+    /// </summary>
+    [Serializable()]
+    public class SimulationException : Exception
+    {
+        private readonly string? source;
+
+        /// <summary>
+        ///     The name of the native shared library which raised this
+        ///     exception if known, <c>null</c> otherwise.
+        /// </summary>
+        public string? SourceLibrary => source;
+
+        internal SimulationException(string message, string? source) : base(message)
+        {
+            this.source = source;
+        }
+    }
+
+    /// <summary>
+    ///     Abstracts away calls to and from the experimental simulators DLL.
+    /// </summary>
+    internal static class NativeInterface
+    {
+        public static event Action<string>? OnVerbose = null;
+        private static void LogCall(string callName) =>
+            OnVerbose?.Invoke($"[VERBOSE] NativeInterface: calling {callName}.");
+
+        private static void CheckCall(Int64 errorCode)
+        {
+            if (errorCode != 0)
+            {
+                var error = _LastError();
+                throw new SimulationException($"Exception in native open systems simulator runtime: {error}", DLL_NAME);
+            }
+        }
+
+
+        public const string DLL_NAME = "Microsoft.Quantum.Experimental.Simulators.Runtime.dll";
```

Comment:
FYI: I assume that the current support covers Windows only.

Reply:
No, this is cross-platform, as is tested by e2e tests in our cross-repo builds. As with the existing `QuantumSimulator` class, however, we use the extension `*.dll` on all platforms when interoperating with .NET to avoid having to write a custom DLL loader that correctly redirects to OS-specific shared library names.